### DISCUSSION TITLE
Работа фильтра dedupTracks

### DISCUSSION THREAD

**Discussion by Bondian** [Chimildic/goofy#116] [Принцип работы]
Фильтр dedupTracks не удаляет дубликаты по имени трека? Или я не понимаю его работу?

```
  let AllTracksOfArtist = Source.getArtistsTracks({
    artist: {
      followed_include: false,
      include: [
        { id: '3xmyGL7g6crcLsRy1TNOQU', name: 'Secret Service' },

      ],
    },
    album: {
      groups: 'album,single', // ....and appears_on,compilation
    },
  });

  Filter.dedupTracks(AllTracksOfArtist);
  Order.sort(AllTracksOfArtist, 'meta.name', 'asc');

  Playlist.saveAsNew({
    name: 'Весь Secret Service',
    public: false,
    tracks: AllTracksOfArtist,
  `});`
```

**Comment by Chimildic** [comment]
Обнаружение дубликата как по _id_, так и по имени, но есть нюансы:
- Не смотря на аргументы запроса, Spotify дает дубликаты альбомов для разных стран.
- Исполнитель мог выпустить переиздание по сути того же трека, но в новом альбоме.
- Треки могут называться логически одинаково, но по написанию отличаться (например  `and` и `&`)
- Треки с разным _id_, одинаковым именем, но разной продолжительностью.

Исполнитель `Secret Service` собрал все нюансы: переиздание, другие написания имен, разная продолжительность. 

В `dedupTracks` есть отклонение в 2 секунды, при котором треки считаются одинаковыми. Обратите внимание на продолжительность треков у переизданий `Secret Service`. Есть примеры с 10 и больше секундами. При этом имея одинаковые названия. Поэтому на выходе много дубликатов по имени с разной продолжительностью.

Увеличив отклонение, остаются лишь дубликаты с разным написанием (как раз `and` и `&`). Первый раз такой исполнитель попался. Все исключения в себе собрал.

Вероятно имеет смысл увеличить и в основном коде, сделаю позднее. Пока вы можете поправить вручную. Нужно найти в файле `library` функцию `isDuplicateByName` и добавить ноль к `2000` (продолжительность в миллисекундах). То есть сделать 20 секунд вместо 2.

**Reply by efremych** [reply]
> Вероятно имеет смысл увеличить и в основном коде, сделаю позднее. Пока вы можете поправить вручную. Нужно найти в файле library функцию isDuplicateByName и добавить ноль к 2000 (продолжительность в миллисекундах). То есть сделать 20 секунд вместо 2.

20 секунд - это уже другой трек. 2 секунды самое то

**Reply by Bondian** [reply]
Спасибо за развёрнутое пояснение. Всё понятно. Хотя... Ещё один вопрос в догонку.
> Исполнитель мог выпустить переиздание по сути того же трека, но в новом альбоме.

> 

То есть, даже если имя и продолжительность трека одинаковая, но разные альбомы, то он уже не распознаётся как дубликат?

**Reply by Chimildic** [reply]
@Bondian сам по себе альбом не важен. Он просто источник появления дубликата. Поскольку без сингла или альбома трека быть не может. Сравнивается _id_ трека, название в комбинации с исполнителем и продолжительность.

**Reply by Bondian** [reply]
Теперь с распознаванием дубликатов всё ясно.
Возник другой вопрос. По какому алгоритму выбираются треки для удаления? Допустим обнаружено 10 дубликатов одного трека. Какой из них останется? Самый свежий или старый по дате релиза? Или выбор идёт по другому принципу?

**Reply by Chimildic** [reply]
@Bondian первая встреча считается оригиналом, все последующее дублями. 
Если нужно играть со старый/новый, примените до фильтра соответствующую сортировку. 
У истории прослушиваний уже сортировка от свежих к старым. Поэтому оригиналом будет недавнее прослушивание.

**Reply by Bondian** [reply]
Спасибо!!!

**Comment by nik1071** [comment]
![image](https://user-images.githubusercontent.com/46973807/166414932-f22bcca5-c10a-4157-8211-b0b0905b56da.png)
Добрый день! Дабы не создавать новую тему, вопрос. Как-то можно такие случаи фильтровать? Или хотя бы самому где-то подправить в Library.gs.

**Reply by Chimildic** [reply]
Попробуйте модифицировать функцию `formatName` из файла _library.gs_. Добавляется строка `.replace(/(feat.+)/, ' ')`. Место вставки также важно.
```js
String.prototype.formatName = function () {
    return this.toLowerCase()
        .replace(/['`,?!@#$%^&*()+-./\\]/g, ' ')
        .replace(/(feat.+)/, ' ')
        .replace(/\s{2,}/g, ' ')
        .replace(/ё/g, 'е')
        .trim();
};
```

**Reply by nik1071** [reply]
Спасибо, убрались с `(feat.)`, но остались с кавычками. Дополнил первую строку с 

> .replace(/["«»<>'`,?!@#$%^&*()+-./\\]/g, ' ')

Вроде все отфильтровалось. Так ведь можно? "Подводных камней" не было б, пытаюсь синхронизировать Spotify и Yandex Music, а из яндекса вечно переносит совершенно с других альбомов те же самые песни.
